### ISSUE TITLE
documentation nit, on one-line function definition syntax

### ISSUE THREAD

**Issue by gwhitney** [DanielXMoore/Civet#723]
The (relatively) new reference page shows three examples of one-line function definitions:
```
function do console.log "Anonymous"
function f do console.log "Named"
function f(x) do console.log x
```
strongly implying the `do` is necessary. But all three examples work perfectly fine without `do`.

**Comment by edemaine** [comment]
This is because it's the recommended way to write single-line functions. Without a do, it can be ambiguous. Consider `function val` vs `function do val`, for example. We might disable the non-do forms... Or at least we haven't decided yet.

**Comment by gwhitney** [comment]
Gotcha. Is there any ambiguity in the case when there's an argument list? If not, then the one-line forms could be
```
function f(x) console.log x
function f() console.log 'Look ma no args!'
-> console.log "Anonymous"
```
(I am a little unclear on what good the anonymous one-line `function do stuff` provides anyway that `->` and/or `=>` don't already cover.)

**Comment by edemaine** [comment]
With a parenthesized argument list, I agree it's pretty reasonable. But the parens are optional; without them, it's ambiguous. Currently Civet will parse as many things as look like arguments as arguments, then switch to body parsing, which is well-defined, but seems dangerous when writing actual code.

**Comment by gwhitney** [comment]
> which is well-defined, but seems dangerous when writing actual code.

Indeed. I guess that's why I have embraced paren-free function calling but left the parens in always when defining functions. I feel like that's a case where the parens really help readability rather than just add useless characters.

So I guess I will just keep writing 'function f() computeStuff' for my one-liners and hope y'all don't disallow that in a future release. To my eyes it reads more nicely than `function f do computeStuff` ... and is one character shorter ;-) ;-P

**Comment by edemaine** [comment]
Actually, I thought `function f x` was shorthand for `function f(x)` but that doesn't exist yet. I forgot that #407 hasn't been resolved yet. 😅 

We could document `function f(x) console.log x` as an explicit alternative, but one more counterargument is that it looks pretty weird with types (`function f(x): void console.log x`). Alternatively, we could say something like "`do` is usually optional but can make things easier to read".

**Comment by gwhitney** [comment]
> pretty weird with types (`function f(x): void console.log x`).

Dunno, I don't think it's any weirder than the types in `(): string => 'foo'` or `{data: number, tail: List} := myList`. And I suspect most one-liners are going to allow their return types to be inferred anyway, so the most likely place you'd be typing is inside the arg list.

However, I do feel like the issues raised in #407 about type annotations constitute fairly strong arguments against parenthesis-optional function declarations. And parentheses are not optional in `=>` and `->` notations, and that seems fine. So it's entirely consistent for them to be required in function declarations. Punctuation was invented to improve readability in natural languages, and it can and does do that in designed languages, especially where it's not redundant with other aspects of the syntax, including layout.

So my concluding 2 cents is 'Embrace parentheses in function definitions consistently, and forget the `do`.'  I think it's OK that function definition and application syntax could differ slightly. Following that principle systematically would say that since `=> executeStuff()` is allowed (no parens in argumentless case), then `function f executeStuff()` should be allowed as well. Which it could well be, with a rule like "The next identifier after the function keyword is the function name." That would disallow the anonymous `function console.log 'bar'` but if you really want an anonymous function like this, why can't you just write `-> console.log 'bar'` -- doesn't that amount to the same thing?

Sorry if this comment should have been _in_ #407, really.
